src/libstore/derivations.cc

```diff
@@ -491,12 +514,15 @@ DrvHashModulo hashDerivationModulo(Store & store, const Derivation & drv, bool m
     }
     case DerivationType::InputAddressed:
         break;
+    case DerivationType::DeferredInputAddressed:
+        break;
```

Comment:
I think we need to set `hasUnknownHash = true` here or something so that `DeferredInputAddressed` cascades to downstream input-addressed derivations?

Reply:
I'm not sure. The reasoning behind the current code is that that way we can call `hashDerivationModulo` on a derivation that's advertising as `DeferredIA` but actually has a statically known hash. It's used in `rewriteDerivation` to convert the `DeferredIA` into an `IA`.
And if the derivation is indeed `Deferred`, then at least one recursive call to `hashDerivationModulo` will return `UnknownHash{}` so we'll get the same result

Reply:
Ah OK. That makes total sense. Then I'd say just one thing we might do is add a flag to `hashDerivationModulo` about whether it should try to un-defer `DeferredIA` derivations: in `build.cc` it definitely should just like this, but I think in `primops.cc` it shouldn't because that be impure, leaking what derivations we've already built into eval itself.

Reply:
> that be impure, leaking what derivations we've already built and which CAs we got from them, into eval itself.

I don't think it does, `hashDerivationModulo` only looks at the derivations themselves. So if we depend on a floating-ca derivation it will always return `UnknownHash{}`, regardless of whether it has been built or not. The reason why we have a different return in `build.cc` is that we call it on the resolved derivation which is still marked as `Deferred` (because we haven't rebuilt its inputs yet) but doesn't depend on any floating-ca derivation anymore.

Reply:
Ah right, and it's only `drvPathResolutions` that we insert into by hand in `build.cc`, whereas `drvHashes` is just manually inserted into in `primops.cc`, where as you say there is nothing intrinsically impure going on. Sorry for the noise then! It's tricky but ends up working out great :).

Reply:
Wait, one thing is that input-addressed derivations conventionally keep their their input-addressed input drvs. But if we rewrite then `hashDerivationModulo`, they won't have any input drvs, even if those input drvs were input-addressed.

This isn't too bad, in fact this is how we "wish" input-addressed derivations worked all along, but it does mean that if one replaced a fixed output derivation with a floating CA derivation that produces the same outputs, one will get a different input-addressed path.